[NativeAOT-LLVM] Add support for GT_STORE_OBJ and GT_OBJ

yowl opened the pull request:
This PR add support for GT_STORE_OBJ and GT_OBJ, including support for the zero initializer when storing a 0 constant.

It implements a call back to Ilc to get the structure layout in one call, and then implements a recursive store similar to what was done in Ilc.

Closes #1757 

cc @SingleAccretion

yowl committed:
enable GT_STORE_OBJ/GT_OBJ

yowl committed:
remove old gtHasHiddenArgument

yowl committed:
remove old gtHasHiddenArgument

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1593,6 +1677,9 @@ void Llvm::visitNode(GenTree* node)
         case GT_STOREIND:
             importStoreInd((GenTreeStoreInd*)node);
             break;
+        case GT_STORE_OBJ:
+            importStoreObj((GenTreeStoreInd*)node);
+            break;
```
```suggestion
        case GT_STOREIND:
            importStoreInd(node->AsStoreInd());
            break;
        case GT_STORE_OBJ:
            importStoreObj(node->AsStoreInd());
            break;
```
(Would also be nice to rename these `buildStoreXYZ` for consistency, or at least the new function)

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter_Statics.cs:
```diff
@@ -379,6 +379,40 @@ public static int PadOffset(TypeDesc type, int atOffset)
+        {
+            foreach (FieldDesc f in type.GetFields())
+            {
+                if (f.IsStatic) continue;
```
```suggestion
                if (f.IsStatic)
                {
                    continue;
                }

```
(Or without the braces if that's the prevailing style)

SingleAccretion commented on src/coreclr/jit/llvm.h:
```diff
@@ -64,6 +64,14 @@ struct SsaPairHash
     }
 };
 
+struct FieldStoreLayout
+{
+    unsigned AbsoluteOffset;
+    unsigned FieldOffset;
+    unsigned IsStartStruct;
+    unsigned IsEndStruct;
+};
```
Hmmm, `FieldStoreLayout`.

This feels more like `FieldSlotDesc` or something to me.

This needs some documentation, on the whole and for the individual members: what sort of fields are recorded, some examples on what the array of these (which is what the "type layout" as a whole will be) looks like for some interesting cases, e. g.
```cs
struct StructWithObj { object Obj; }
struct NestedStructWithObj { StructWithObj StructWithObj; }
struct StructNoGcPtr { int A, B, C, D; }
struct StructComposite { int A; object Obj; }
struct StructMixed { int A; object Obj, double C; StructWithObj StructWithObj; }
```
Also, I wonder if it would have been fruitful to try and pass the full representation of the VM layout to the Jit (build a small copy of `FieldDesc` in the Jit). Seems like it could've been shared with the "build LLVM struct type" code.

yowl replied:
I suppose this goes in hand with not double copying the memory below.  I think the problem I'm having is how to create a recursive struct in the C# side, that I can marshal to the C++ side.  I'll need that wont I?

SingleAccretion replied:
Yes, but it does not have to be recursive on the interface side. E. g.:
```c
struct TypeDescriptor
{
     size_t fieldCount;
     CORINFO_FIELD_HANDLE* fields;
}

TypeDescriptor getTypeDescriptor(CORINFO_CLASS_HANDLE clsHnd);
```
Then we should be able to build an appropriate recursive abstraction on the Jit side.

yowl replied:
Thanks.  What I'll do then is create another PR for just this and incorporate the new struct into the build LLVM struct type code.  Then I'll come back to this one.

SingleAccretion commented on src/coreclr/jit/llvm.h:
```diff
@@ -133,6 +142,7 @@ class Llvm
     void buildHelperFuncCall(GenTreeCall* call);
     llvm::FunctionType* buildHelperLlvmFunctionType(GenTreeCall* call, bool withShadowStack);
     void buildInd(GenTree* node, Value* ptr);
+    void buildObj(GenTreeObj* node, Value* ptr);
```
Nit: `void buildObj(GenTreeObj* node, Value* ptr)` -> `void buildObj(GenTreeObj* obj)` and retrieve the pointer inside it.

(Realize `buildInd` is doing this, it should be fixed too -- possibly as a separate cleanup)

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs:
```diff
@@ -287,6 +288,31 @@ public static CorInfoTypeWithMod getParameterType(IntPtr thisHandle, CORINFO_CLA
+            uint fieldCount = (uint)layout.Count;
+
+            // TODO-LLVM: release?  Is there a smarter way of doing this
+            var layoutPtr = Marshal.AllocHGlobal((int)(sizeof(FieldStoreLayout) * fieldCount));
```
Does the compiler project reference .NET ref recent enough for it to have `NativeMemory.Alloc`? Would be nice to use (if possible) it as it's much faster than `AllocHGlobal`.

yowl replied:
No, this project is .net 5.0 unfortunately.  We are some way behind `main`, if we caught up we could.  I'll put a todo for now.

yowl replied:
Or we could just change it, and worry about merging later.

SingleAccretion replied:
Up to you, it's a little thing, may not be worth the hassle of updating to 6.

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs:
```diff
@@ -287,6 +288,31 @@ public static CorInfoTypeWithMod getParameterType(IntPtr thisHandle, CORINFO_CLA
+
+            uint fieldCount = (uint)layout.Count;
+
+            // TODO-LLVM: release?  Is there a smarter way of doing this
```
This does seem like it would leak memory globally for the process, doesn't it.

Presumably there is a "cleanup once the compilation is over" mechanism in this Jit-EE interface implementation, and presumably we should be using it here to release the memory.

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs:
```diff
@@ -287,6 +288,31 @@ public static CorInfoTypeWithMod getParameterType(IntPtr thisHandle, CORINFO_CLA
+
+            TypeDesc type = _this.HandleToObject(inputType);
+
+            List<FieldStoreLayout> layout = new();
+
+            _this._compilation.GetObjectLayoutInstructions(type, layout);
```
Nit: cleaner to return the list from `GetObjectLayoutInstructions` instead of mutating it?

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs:
```diff
@@ -287,6 +288,31 @@ public static CorInfoTypeWithMod getParameterType(IntPtr thisHandle, CORINFO_CLA
+
+            for (int i = 0; i < fieldCount; i++)
+            {
+                Marshal.StructureToPtr(layout[i], layoutPtr + sizeof(FieldStoreLayout) * i, false);
```
Isn't this just:
```cs
((FieldStoreLayout*)layoutPtr)[i] = layout[i];
```
?

(Also would be nice to get rid of the casts by typing `layoutPtr` as `FieldStoreLayout` and `layoutPtrPtr` as `FieldStoreLayout**`)

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1151,6 +1157,13 @@ void Llvm::buildInd(GenTree* node, Value* ptr)
+{
+    // cast the pointer to create the correct load instructions
+    mapGenTreeToValue(node, _builder.CreateLoad(
+                                castIfNecessary(ptr, getLlvmTypeForStruct(node->GetLayout()->GetClassHandle())->getPointerTo())));
```
I think we'd want to make an overload of `getLlvmTypeForStruct` that takes a `ClassLayout*` directly.

That way we'll be able to deal with "block layouts" (that do not have a struct handle) transparently in the future.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+        {
+            if (data->getType()->isStructTy())
+            {
+                unsigned index = _module->getDataLayout().getStructLayout((llvm::StructType*)(data->getType()))->getElementContainingOffset(gcFieldLayout[i].FieldOffset);
```
This could use an intermediate variable for the struct layout.

(Also, `index` -> `llvmFieldIndex`)

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+    FieldStoreLayout* gcFieldLayout;
+    gcFieldCount = _getObjectLayout(_thisPtr, structClsHnd, &gcFieldLayout);
+
+    storeObjAtAddress(baseAddress, getGenTreeValue(storeIndOp->Data()), 0, gcFieldCount, gcFieldLayout);
+
+    // just copy all the fields again for simplicity,
+    // if all the fields were set using RhpAssignRef then a possible optimisation would be to skip this line
+    Value* toStore = getGenTreeValue(storeIndOp->Data());
+    _builder.CreateStore(toStore, castIfNecessary(baseAddress, toStore->getType()->getPointerTo()));
```
Hmm, it does not seem right to me to store to the same location twice.

Presumably it breaks the GC to store a ref to the heap without a barrier, and the second store would be doing that.

Also, why do we need to do it all? Wouldn't the equivalent of the below (that looks a lot like the current IL code):
```cs
void StoreObjFields(addr, offset, obj)
{
    if (obj.IsGCPtr()) // TYP_REF/TYP_BYREF
    {
        StoreWithWriteBarrier(addr + offset, obj);
        return;
    }
    if (!obj.HasGCPtr())
    {
        StoreLlvm(addr + offset, obj);
        return;
    }

    foreach (field in obj)
    {
        StoreObjFields(addr, offset, field);
        offset += field.Size; // With padding
    }
}
```
Store the whole struct (modulo padding -- we'll need to handle it with `memcpy` for explicitly-sized structs)?

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+    {
+        _builder.CreateMemSet(baseAddress, _builder.getInt8(0), _builder.getInt32(structLayout->GetSize()), {});
+        return;
+    }
```
This needs a `TODO-LLVM` for `handle non-zero InitBlks.`.

(These are generated for the `initblk` IL instruction that has the same semantics as `memset`. In IR they will be block stores with `INIT_VAL` `data`)

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+
+    // zero initialization  check
+    GenTree* dataOp = storeIndOp->Data();
+    if (dataOp->OperIs(GT_CNS_INT) && dataOp->AsIntCon()->IconValue() == 0)
```
```suggestion
    if (dataOp->IsIntegralConst(0))
```

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+
+    CORINFO_CLASS_HANDLE structClsHnd = structLayout->GetClassHandle();
+    FieldStoreLayout* gcFieldLayout;
+    gcFieldCount = _getObjectLayout(_thisPtr, structClsHnd, &gcFieldLayout);
```
Re-creating this for each node seems expensive. We need a side table that'll associate it with `ClassLayout*` pointers.

SingleAccretion reviewed (commented):
Thinking we should probably try and see how replicating the full field information on the Jit side is going to look like, seems we need it.

yowl committed:
rename import->build

yowl committed:
Address feedback

yowl committed:
Merge remote-tracking branch 'origin/feature/NativeAOT-LLVM' into store-obj

yowl committed:
Use StructDesc and FieldDesc for GT_STORE_OBJ

yowl commented:
@SingleAccretion The change from #1854 is now incorporated here, removing the "instruction" struct.  I've not needed to make it recursive in the end.

It's passing the tests, but I have 2 outstanding questions

1. How can I identify whether a field in the struct needs an `RhpAssignRef`?  Right now I just have
```
if (fieldClassHandle != NO_CLASS_HANDLE)
{
    isGcPtr = corInfoType == CORINFO_TYPE_CLASS;  //TODO-LLVM: how to tell if the field needs an RhpAssignRef - does _getTypeDescriptor have to do that?
}
```
2. Should I be filling in spaces between fields with 0s ?

SingleAccretion commented:
> How can I identify whether a field in the struct needs an `RhpAssignRef`?

`isGcPtr = corInfoType == CORINFO_TYPE_CLASS` looks right, but the existing backend code also does this for `BYREF`s (see `genCodeForCpObj` in `codegenxarch.cpp`), but I do not sure why, given byrefs shouldn't be living on the heap. You can (should) also omit the write barrier if the destination address is local (or `TGT_NOT_HEAP`).

> Should I be filling in spaces between fields with 0s ?

In RyuJIt we assume padding is not significant in case it is "natural", that is, the struct is not `LayoutKind.Explicit` and doesn't have the `Size` metadata.

So if the struct "has custom layout" like that, it'll need to be copied wholesale, if not - it can be copied field by field, ignoring padding. In either case, you don't need to zero anything.

yowl committed:
Dont use write barrier if GTF_IND_TGT_NOT_HEAP

include BYREF as GCPtr test

yowl commented:
I'll write some tests for `ExplicitLayout`

yowl committed:
add padding in copy if struct explicit

Use memcpy if targetNotHeap

yowl commented:
I added some code to add an explicitlayout flag to TypeDescriptor so that a decision can be made if copying of gaps is required.  I added a test for that, but it doesn't compile with the clrjit module yet.

If the target is not heap, then it will just do a memcpy.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
     }
 }
 
+// Copies endOffset - startOffset bytes, endOffset is exclusive.
+unsigned Llvm::buildMemCpy(Value* baseAddress, unsigned startOffset, unsigned endOffset, Value* srcAddress)
+{
+    Value* paddingDestAddress = _builder.CreateGEP(baseAddress, _builder.getInt32(startOffset));
+
+    const unsigned paddingSize = endOffset - startOffset;
+    // copy the bytes in the padding
+    _builder.CreateMemCpy(paddingDestAddress, llvm::Align(), srcAddress, llvm::Align(), paddingSize);
+
+    return paddingSize;
+}
```
```suggestion
// Copies endOffset - startOffset bytes, endOffset is exclusive.
unsigned Llvm::buildMemCpy(Value* baseAddress, unsigned startOffset, unsigned endOffset, Value* srcAddress)
{
    Value* destAddress = _builder.CreateGEP(baseAddress, _builder.getInt32(startOffset));
    unsigned size = endOffset - startOffset;

    _builder.CreateMemCpy(destAddress, llvm::Align(), srcAddress, llvm::Align(), size);

    return size;
}
```
Doesn't seem like there is great a reason to make this padding-copying-specific.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+            // recurse into struct
+            storeObjAtAddress(address, fieldData, getStructDesc(fieldDesc->getClassHandle()));
+
+            bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / 8;
```
```suggestion
            bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / BITS_PER_BYTE;
```

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+            {
+                _builder.CreateStore(fieldData, castIfNecessary(address, fieldData->getType()->getPointerTo()));
+
+                bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / 8;
```
```suggestion
                bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / BITS_PER_BYTE;
```

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+        }
+    }
+
+    unsigned llvmStructSize = data->getType()->getPrimitiveSizeInBits() / 8;
```
```suggestion
    unsigned llvmStructSize = data->getType()->getPrimitiveSizeInBits() / BITS_PER_BYTE;
```

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+
+    CORINFO_CLASS_HANDLE structClsHnd  = structLayout->GetClassHandle();
+    StructDesc*          structDesc    = getStructDesc(structClsHnd);
+    bool                 targetNotHeap = (indirOp->gtFlags & GTF_IND_TGT_NOT_HEAP) == 0;
```
```suggestion
    bool                 targetNotHeap = ((indirOp->gtFlags & GTF_IND_TGT_NOT_HEAP) != 0) || genTreeObj->Addr()->OperIsLocalAddr();
```
The condition seems to have been inverted.

yowl replied:
oops

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+    }
+}
+
+void Llvm::buildStoreObj(GenTreeIndir* indirOp)
```
The method is named `buildStoreObj` but takes `GenTreeIndir*`. Why not make it take `GenTreeObj*`?

(Also I would rename `indirOp` to something with `store` in the name)

I think we need to establish a naming convention for trees-vs-LLVM-values, they are mixed right now and it can be confusing. Appending `node` for trees seems a fine solution, e. g. here we could have:
```c
GenTree* dstAddrNode = storeObjNode->Addr();
Value* baseAddress = getGenTreeValue(dstAddrNode);
```
Or we could leave the trees with "simple" names and append `Value` to `llvm::Value*`s instead, e. g.:
```
GenTree* dstAddr = storeObj->Addr();
Value* baseAddressValue = getGenTreeValue(dstAddr);
```
Or have it be as it is (simple names for both).

What do you think?

SingleAccretion replied:
@yowl ^

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+
+void Llvm::buildStoreObj(GenTreeIndir* indirOp)
+{
+    Value* baseAddress = getGenTreeValue(indirOp->Addr());
```
This should be moved down to where it is first used.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+    StructDesc*          structDesc    = getStructDesc(structClsHnd);
+    bool                 targetNotHeap = (indirOp->gtFlags & GTF_IND_TGT_NOT_HEAP) == 0;
+
+    Value* toStore = getGenTreeValue(genTreeObj->Data());
```
`toStore` -> `dataValue`.

SingleAccretion commented on src/coreclr/jit/llvm_types.h:
```diff
@@ -47,18 +70,24 @@ struct FieldDesc
         return m_corType;
     }
+
+    bool isGcPointer()
+    {
+        return m_isGcPointer;
+    }
```
```suggestion
    bool isGcPointer()
    {
        return m_corType == CORINFO_TYPE_CLASS || m_corType == CORINFO_TYPE_BYREF;
    }
```
Seems like there's no need to cache it.

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs:
```diff
@@ -295,6 +295,7 @@ public struct TypeDescriptor
         {
             public uint FieldCount;
             public CORINFO_FIELD_STRUCT_** Fields; // array of CORINFO_FIELD_STRUCT_*
+            public uint IsExplicitLayout;
```
We should be mindful not to waste too much space here, since it is cached (this is 32 bits for a boolean). It's fine as-is right now, but if we add more booleans like this, we should make this `uint` a flags value.

SingleAccretion commented on src/coreclr/jit/llvm_types.h:
```diff
@@ -12,8 +12,30 @@
+        return m_fields[index];
+    }
+
+    unsigned getIsExplicitLayout()
+    {
+        return m_isExplicitLayout;
+    }
```
The condition here is: `LayoutKind.Explicit || HasSizeMetadata`, i. e. the following struct should also be treated as one with significant padding:
```cs
[StructLayout(LayoutKind.Sequential, Size = 10)]
struct S
{
    public byte Field;
}
```
This would also imply renaming this (and related interface code) to something like `hasSignificantPadding`.

yowl replied:
Thanks

SingleAccretion commented on src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter_Statics.cs:
```diff
@@ -379,6 +379,44 @@ public static int PadOffset(TypeDesc type, int atOffset)
             return padding;
         }
 
+        public static void GetObjectLayoutInstructions(TypeDesc type, List<FieldStoreLayout> layout, int startOffset)
```
This and related code unused?

yowl committed:
Update src/coreclr/jit/llvm.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/llvm.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Apply suggestions from code review

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
address feedback.  Cleanup dead code.

yowl committed:
Merge branch 'store-obj' of https://github.com/yowl/runtimelab into store-obj

yowl committed:
sequential layout with size

yowl committed:
Update src/coreclr/jit/llvm.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl commented:
I think I've done all the review items now.

yowl committed:
feedback, naming improvements

yowl committed:
Merge branch 'store-obj' of https://github.com/yowl/runtimelab into store-obj

yowl committed:
move var closer to use

jkotas reviewed (approved):
Thank you!
